### ISSUE TITLE
Feature Request: Rework button graphics generation

### ISSUE THREAD

**Issue by dnmeid** [bitfocus/companion#262]
1. Let's unify the button styles further. In my opinion a button should have layers, a color background layer, a png layer and a text layer. So we don't need two different button styles anymore. And we could combine background colors with transparent pngs. And maybe for the future we can offer the ability to let the user add more layers or give more options to layers or more types of layers.
2. There should be an option to resize pngs and alignment control option. Resizing should not be done during upload, but during presentation.
3. Feedbacks could also use one or more layers, let's say a feedback is writing some text into a text layer and the user can have another textlayer writing his own text.
3. Button states. Let's give each button the ability to have more than one design. This would make much things much easier. Each state has all the options of the button design in it, so a module or the user can prepare many states and switch between them only with changing the state. Great for toggle with state 0 and state 1. But there could be more states, feedback can also be used to switch states, so it is easy to implement feedback and at the same time it is comprehensible to the user and easy adjustable.

**Comment by haakonnessjoen** [comment]
I also want to add support for uploading gifs. But with some limitations. They will always loop, and a fixed delay between each frame. So a gobal timer for all gifs.

It should actually be very simple to implement, with js-only modules, so I might, some day ;P

**Comment by edifyyo** [comment]
I'd very much like GIF support, as well!

**Comment by RobbieBott** [comment]
+1 for looping gif support.

**Comment by Julusian** [comment]
I think that supporting animated gifs will require some technical thought about how to avoid it becoming a performance hog.

* Ideally we should be able to generate each frame once and to reuse it upon looping. How many frames is too many? we don't want to use too much ram to buffer these. 
* do we lazy render them, or render all frames at once? lazily will cause some more complex logic for the display side, but all at once will cause burstier cpu and be wasteful in the case of long gifs and fast changing feedbacks.
* how do we efficiently transfer these renders to connected clients/devices? it would be good to avoid resending the same data repeatedly, this becomes a similar problem to the rendering
* mixed framerates should be ok, if the frame ticking is done outside the core (clients/usb process)
* what will be the performance impact of this? with companion being mostly single threaded I would be concerned about the impact that the extra rendering and data shuffling will have on device operations. This might be something to put off until there is more threading going on

**Comment by barnes2000** [comment]
I would very much like to see animated gifs added to companion for stream deck. I am not an engineer, but if there are donations needed to help expedite this, please let me know. Thank you.

**Comment by barnes2000** [comment]
It's 2022 now and I'd like to bump this request again. I am willing to donate $$$ to help expedite this feature, if there's someone that thinks they can achieve it. Thank you.

**Comment by ukiews** [comment]
Is there a reason something simple like this cannot be implemented? Currently this is the only drawback that requires us to create buttons within streamdeck software and run companion as a plugin. For example, when OBS is livestreaming, we want to be able to have the button blink "LIVE"

**Comment by thedist** [comment]
> Is there a reason something simple like this cannot be implemented? Currently this is the only drawback that requires us to create buttons within streamdeck software and run companion as a plugin. For example, when OBS is livestreaming, we want to be able to have the button blink "LIVE"

Blinking text/colours is all already possible in Companion.

**Comment by ukiews** [comment]
> > Is there a reason something simple like this cannot be implemented? Currently this is the only drawback that requires us to create buttons within streamdeck software and run companion as a plugin. For example, when OBS is livestreaming, we want to be able to have the button blink "LIVE"
> 
> Blinking text/colours is all already possible in Companion.

I suppose this could be accomplished by using variables and delays, but this would be unnecessarily involved, and I imagine prone to issues that would hug up resources.

Could we not get a simple option to add a GIF like streamdeck already allows?

**Comment by thedist** [comment]
> I suppose this could be accomplished by using variables and delays, but this would be unnecessarily involved, and I imagine prone to issues that would hug up resources.

Or just use the `generic-blink` module.

**Comment by markmagoo** [comment]
Just curious if APNGs would make a difference here? Since they're already using PNGs, is that too far a reach?

**Comment by dnmeid** [comment]
> Just curious if APNGs would make a difference here? Since they're already using PNGs, is that too far a reach?

No, it is not a matter of the format. GIF, APNG, animated SVG, decoding would be possible with small effort. The problem is the performance on one hand and someone who has the time to work at it. Julian already explained it, there will be a lot of code to write, fix and improve.
